File: 28682-FENTREE/main.go
package main

import "fmt"

// Fenwick tree
type fentree struct {
	tree []int64
}

func build(arr []int64) *fentree {
	numElements := len(arr)

	t := fentree{}
	t.tree = make([]int64, numElements)

	for i := 0; i < numElements; i++ {
		t.update(i, arr[i])
	}

	return &t
}

func (t *fentree) update(index int, value int64) {
	n := len(t.tree)
	for i := index; i < n; i |= i + 1 {
		t.tree[i] += value
	}
}

func (t *fentree) query(left int, right int) int64 {
	return t.sum(right) - t.sum(left-1)
}

func (t *fentree) sum(index int) int64 {
	var res int64
	for index > 0 {
		res += t.tree[index]
		index &= index + 1
		index = index - 1
	}
	return res
}

func main() {
	var numElements int
	fmt.Scan(&numElements)
	arr := make([]int64, numElements+1)
	for i := 1; i <= numElements; i++ {
		fmt.Scan(&arr[i])
	}

	tree := build(arr)

	var numOperations int
	fmt.Scan(&numOperations)

	var operation string
	var a, b int
	for i := 0; i < numOperations; i++ {
		fmt.Scan(&operation, &a, &b)
		if operation == "q" {
			fmt.Println(tree.query(a, b))
		} else {
			tree.update(a, int64(b))
		}
	}
}

File: 28682-FENTREE/main.cpp
#include <iostream>
#include <vector>
using namespace std;

struct FenwickTree
{
    vector<long long> elements;

    void buildFrom(vector<int>& arr)
    {
        elements.resize(arr.size(), 0LL);
        for(size_t i = 0; i < arr.size(); i++) {
            update(i, arr[i]);
        }
    }

    void update(int index, int value)
    {
        size_t n = elements.size();
        for(size_t i = index; i < n; i = i | (i+1)) {
            elements[i] += value;
        }
    }

    long long query(int left, int right)
    {
        return sum(right) - sum(left-1);
    }

    long long sum(int index)
    {
        long long res = 0LL;
        while(index > 0)
        {
            res += elements[index];
            index = index & (index + 1);
            index--;
        }
        return res;
    }
};

int main()
{
   int numElements;
   cin >> numElements;

   vector<int> arr;
   arr.push_back(0LL);
   int n = 0;
   for(int i = 1; i <= numElements; i++) {
       cin >> n;
       arr.push_back(n);
   }

    FenwickTree tree;
    tree.buildFrom(arr);

   int numOperations;
   cin >> numOperations;

   char operation = 0;
   int a = 0;
   int b = 0;

    for(int i = 0; i < numOperations; i++) {
        cin >> operation;
        cin >> a;
        cin >> b;

        if(operation == 'q') {
            cout << tree.query(a, b) << endl;
        } else {
            tree.update(a, b);
        }
   }

    return 0;
}